Public/Src/Tools/Execution.Analyzer/Analyzers.Core/XLGPlusPlus/XldbDataStore.cs

```diff
@@ -104,80 +117,270 @@ private IEnumerable<string> GetEventsByType(ExecutionEventId eventTypeID)
             return storedEvents;
         }
 
+        public IMessage GetEventByKey(ExecutionEventId eventTypeID, uint pipID)
+        {
+            Contract.Assert(Accessor != null, "XldbStore must be initialized via OpenDatastore first");
+            
+            var eventQuery = new EventTypeQuery
+            {
+                EventTypeID = eventTypeID,
+                PipId = pipID
+            };
+
+            var maybeFound = Accessor.Use(database =>
+            {
+                foreach (var kvp in database.PrefixSearch(eventQuery.ToByteArray(), EventColumnFamilyName))
+                {
+                    if (m_eventParserDictionary.TryGetValue(eventTypeID, out var parser))
+                    {
+                        return parser.ParseFrom(kvp.Value);
+                    }
+                }
+                return null;
+            });
+
+            if (!maybeFound.Succeeded)
+            {
+                maybeFound.Failure.Throw();
+            }
+
+            return maybeFound.Result;
+        }
+
+        /// <summary>
+        /// Returns the count and payload of events by the event type
+        /// </summary>
+        /// <returns>EventCountsByTypeValue if exists, null otherwise</returns>
+        public EventCountByTypeValue GetCountByEvent(ExecutionEventId eventTypeID)
+        {
+            Contract.Assert(Accessor != null, "XldbStore must be initialized via OpenDatastore first");
+
+            var eventCountQuery = new EventCountByTypeKey
+            {
+                EventTypeID = eventTypeID,
+            };
+
+            var maybeFound = Accessor.Use(database =>
+            {
+                if (database.TryGetValue(eventCountQuery.ToByteArray(), out var eventCountObj))
+                {
+                    return EventCountByTypeValue.Parser.ParseFrom(eventCountObj);
+                }
+                return null;
+            });
+
+            if (!maybeFound.Succeeded)
+            {
+                maybeFound.Failure.Throw();
+            }
+
+            return maybeFound.Result;
+        }
+
+        /// <summary>
+        /// Gets the total number of stored xlg events in the database, 0 if the accessor was unsuccesful.
+        /// </summary>
+        public uint GetTotalEventCount()
+        {
+            Contract.Assert(Accessor != null, "XldbStore must be initialized via OpenDatastore first");
+
+            var maybeFound = Accessor.Use(database =>
+            {
+                database.TryGetValue(Encoding.ASCII.GetBytes(EventCountKey), out var eventCountObj);
+                return EventCount.Parser.ParseFrom(eventCountObj).Value;
+            });
+
+            if (!maybeFound.Succeeded)
+            {
+                maybeFound.Failure.Throw();
+            }
+
+            return maybeFound.Result;

```

Comment:
Not needed now, but I suspect we are going to get a bunch more point queries or list queries.

Perhaps it would be usefull to create helper methods that contain this logic and the serialization and error handling so we don't have to do it over and over i.e.

int QueryInt32<T>(T message) where T : IMessage { ... }

and/or TryQueryInt32

Reply:
Maybe worth doing this PR as I see `GetPipTableMetaData` :)

---
In reply to: [314812200](https://github.com/microsoft/BuildXL/pull/724#discussion_r314812200) [](ancestors = 314812200)

Reply:
Talked offline -> moving to separate PR.

---
In reply to: [314812365](https://github.com/microsoft/BuildXL/pull/724#discussion_r314812365) [](ancestors = 314812365,314812200)